src/webgpu/shader/execution/robust_access_vertex.spec.ts

```diff
@@ -44,7 +44,430 @@ Vertex buffer contents could be randomized to prevent the case where a previous
+        for (let j = 0; j < slotsPerBuffer; j++) {
+          layoutStr += `layout(location=${currSlot}) in ${p.type} a_${currSlot};\n`;
+          attributeNames.push(`a_${currSlot}`);
+          currSlot++;
+        }
+      }
+    }
+
+    // Vertex buffer descriptors
+    const vertexBuffers: GPUVertexBufferLayoutDescriptor[] = [];
+    {
+      let currSlot = 0;
+      for (let i = 0; i < bufferContents.length; i++) {
+        vertexBuffers.push({
+          arrayStride: slotsPerBuffer * typeInfo.size,
+          stepMode: i === 0 ? 'instance' : 'vertex',
+          attributes: Array(slotsPerBuffer)
+            .fill(0)
+            .map((_, i) => ({
+              shaderLocation: currSlot++,
+              offset: i * typeInfo.size,
+              format: typeInfo.format,
+            })),
+        });
+      }
+    }
+
+    // Offset the range checks for gl_VertexIndex in the shader if we use BaseVertex
+    let vertexIndexOffset = 0;
+    if (p.drawCallTestParameter === DrawCallParameter.BaseVertex) {
+      vertexIndexOffset += p.errorScale;
+    }
+
+    // Construct pipeline that outputs a green fragment, only if we notice any invalid values
+    const vertexModule = t.makeShaderModule('vertex', {
+      glsl: `
+      #version 450
+      ${layoutStr}
+
+      bool valid(float f) {
+        return ${validValues.map(v => `f == ${v}`).join(' || ')};
+      }
+
+      bool validationFunc(${p.type} v) {
+        ${typeInfo.validationFunc}
+      }
+
+      void main() {
+        bool attributesInBounds = ${attributeNames.map(a => `validationFunc(${a})`).join(' && ')};
+        bool indexInBounds = gl_VertexIndex == 0 || (gl_VertexIndex >= ${vertexIndexOffset} &&
+          gl_VertexIndex < ${vertexIndexOffset + numVertices});
+
+        if (attributesInBounds && (${!p.indexed} || indexInBounds)) {
+          // Success case, move the vertex out of the screen
+          gl_Position = vec4(-1.0, 0.0, 0.0, 1.0);
+        } else {
+          // Failure case, move the vertex inside the screen
+          gl_Position = vec4(0.0, 0.0, 0.0, 1.0);
+        }
+      }
+    `,
+    });
+
+    const fragmentModule = t.makeShaderModule('fragment', {
+      glsl: `
+      #version 450
+      precision mediump float;
+
+      layout(location = 0) out vec4 fragColor;
+
+      void main() {
+        fragColor = vec4(0.0, 1.0, 0.0, 1.0);
+      }
+    `,
+    });
+
+    // Pipeline setup, texture setup
+    const colorAttachment = t.device.createTexture({
+      format: 'rgba8unorm',
+      size: { width: 1, height: 1, depth: 1 },
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.OUTPUT_ATTACHMENT,
+    });
+    const colorAttachmentView = colorAttachment.createView();
+
+    const pl = t.device.createPipelineLayout({ bindGroupLayouts: [] });
+    const pipeline = t.device.createRenderPipeline({
+      vertexStage: { module: vertexModule, entryPoint: 'main' },
+      fragmentStage: { module: fragmentModule, entryPoint: 'main' },
+      layout: pl,
```

Comment:
nit: pipeline layout is optional now
```suggestion
    const pipeline = t.device.createRenderPipeline({
      vertexStage: { module: vertexModule, entryPoint: 'main' },
      fragmentStage: { module: fragmentModule, entryPoint: 'main' },
```

Reply:
Done